WcaOnRails/app/controllers/results_controller.rb

```diff
@@ -0,0 +1,279 @@
+# frozen_string_literal: true
+
+class ResultsController < ApplicationController
+  def index
+    support_old_links!
+
+    # Default params
+    params[:region] ||= "world"
+    params[:years] ||= "all years"
+    params[:show] ||= "100 persons"
+
+    @event_param = params[:event_id]
+    @region_param = params[:region]
+    @type_param = params[:type]
+    years_param = params[:years]
+    show_param = params[:show]
+
+    @years = Competition.non_future_years
+
+    @quantities = ["100", "1000"]
+
+    @types = ["single", "average"]
+    @is_average = @type_param == @types[1]
+    value = @is_average ? "average" : "best"
+    capitalized_type_param = @type_param.capitalize
+
+    event_condition = "AND eventId = '#{@event_param}'"
+    event_condition_without_and = "eventId = '#{@event_param}'"
```

Comment:
I'll wait for further input on how to take care of this properly :slightly_smiling_face:

Reply:
I think there are two related, but different things to consider:

1. making sure that the user provided event_param is a valid event id.
2. making sure that the thing we put in the sql query cannot possibly do any harm.

Right now, doing 1) is "good enough", because our `eventId`s are "sql safe" (they don't contain any special characters). That said, I would rest easier if we *also* have the event param go through some sort of sanitization.

1) should be pretty straightforward to do

As for how to do 2)... It looks like [exec_query takes an optional `binds` parameter](https://api.rubyonrails.org/classes/ActiveRecord/ConnectionAdapters/DatabaseStatements.html#method-i-exec_query), I think you can use that to replace `?`s in your SQL string with safely interpolated values. I'm having trouble finding decent documentation on this, unfortunately. It seems like the functionality in part depends on your database driver (in our case, that's the `mysql2` gem).

Urg, I just poked around for a while and couldn't figure out how to get the binds parameter to do anything useful. I eventually stumbled across this post: https://github.com/rails/rails/issues/20498, and it sounds like this just won't work with ActiveRecord + mysql2...

I'm not sure what the best approach is here. Some ideas for future investigation:

1. Find an escape method somewhere and try to remember to call it for every bit of string interpolation we do.
   - Pros: Straightforward
   - Cons: Need to find this method somewhere. Hopefully there's on buried in ActiveRecord somewhere. Also need to remember to use this everytime we do some string interpolation, which is just asking for trouble.
2. Use the `mysql2` gem directly, rather than going through ActiveRecord. `mysql2` looks like it supports [prepared statements](https://github.com/brianmario/mysql2#usage), which should let us do this in a safe way.
   - Cons: This ties us more closely to `mysql2`, which would make moving to a different database adapter (or even a different database!) more difficult in the future.
3. I don't know if this is possible, but: try to build an actual ActiveRecord model for this query. I get that we're not querying from a real table, but maybe AR supports querying from a subselect? If not, maybe we could trick AR by setting up a mysql view?

Of all these options, 3) feels like the most work, but also possibly the best approach. We've had some luck with using mysql views to trick AR before (see the `rails_persons` view for an example). Definitely curious what @jonatanklosko and @viroulep think, though.

Reply:
Just like `competition_from_params` do a "find" on the competition model, I'd just do `Event.c_find!(...)` to make sure the page gives a 500 on an invalid event id.

Reply:
I've added the `Event.c_find!(params[:event_id])` to make sure the page crashes on an invalid `:event_id`. I'm not sure if something like this is necessary for other params too?

- `:type`:
  - it's used to get `value`, which will be either `best` or `average` (there's no other option,) which then is interpolated on some of the queries
  - it's capitalized and interpolated on some of the queries between `Concise` and `Results`, without spaces
- `:region`:
  - it's used to either find a continent or a country. If none is found, no interpolation is made.
- `:show`:
  - it's used to decide which query to run
  - it's splitted and the first part is directly interpolated on all but one query after a `LIMIT `
- `:years`
  - it's splitted; its first part is used to decide if an `=` or `<=` sign will be interpolated on the queries together with the splitted second part, after a `AND year `. If it doesn't match exactly, no interpolation happens.

Do you think some of them need to be (further) sanitized?

Reply:
So looks like `year` and `show` are not really sanitized, would make sense to just parse them to ints.

Reply:
I've added a `.to_i` after splitting them.

Reply:
> `:type` ... will be either best or average (there's no other option,)

As I'm reading the code, it looks like `type` comes directly from the user, gets capitalized, and then injected directly into a SQL string. Why do you say it will be either "best" or "average"?

I agree that the use of `:region` and show `:show` look ok as they are now. Adding a `.to_i` for `show` seems good.

The current use of `:years` does *not* look safe to me. Adding a `.to_i` does feel like it would make things safe.

Reply:
> As I'm reading the code, it looks like type comes directly from the user, gets capitalized, and then injected directly into a SQL string. Why do you say it will be either "best" or "average"?

`type` is used for two things (note the two sub-points on my previous comment.) The first time it's used is [here](https://github.com/thewca/worldcubeassociation.org/pull/4394/files#diff-09cacc147fbd254009efaa431d0cc8b2R17-R18), hence that comment that you quoted. The second one is when it's directly interpolated. Any suggestion on how to improve that?

> The current use of :years does not look safe to me. Adding a .to_i does feel like it would make things safe.

I had already added it, see [here](https://github.com/thewca/worldcubeassociation.org/pull/4394/files#diff-09cacc147fbd254009efaa431d0cc8b2R47)

Reply:
Maybe something like:

```
if !@types.include?(params[:type])
  flash[:danger] = t(".unknown_type")
  return redirect_to results_rankings_path
end
```

would work for the `type` param sanitization?